route_info/src/map_matching.rs

```diff
@@ -0,0 +1,81 @@
+use anyhow::Result;
+use geom::{Distance, LonLat, Pt2D};
+use osm2streets::{Direction, IntersectionID, RoadID};
+use petgraph::prelude::DiGraphMap;
+
+use super::{RawRouteWaypoint, RouteInfo, Waypoint};
+
+impl RouteInfo {
+    /// Extract the endpoints of a LineString in map-space.
+    pub(crate) fn linestring_endpoints(&self, feature: &geojson::Feature) -> Result<(Pt2D, Pt2D)> {
+        // TODO Duplicate non-adjacent points; make a more permissive version of PolyLine.
+        //let polyline = PolyLine::from_geojson(feature, Some(&self.network.gps_bounds))?;
+
+        if let Some(geojson::Geometry {
+            value: geojson::Value::LineString(ref pts),
+            ..
```

Comment:
I think I understand `if let` but can you run me through this @dabreegster? And also the `..`?

Reply:
This is how you unpack the `pts` from a GeoJSON feature I think

Reply:
We have a [geojson::Feature](https://docs.rs/geojson/latest/geojson/struct.Feature.html) struct. It's got a `geometry: Option<Geometry>` field. Inside there is a [value enum](https://docs.rs/geojson/latest/geojson/enum.Value.html) with a bunch of cases. The `if let` is like Rust's [match](https://doc.rust-lang.org/rust-by-example/flow_control/match.html) statement, but for handling just one case. So we're saying "if this GJ feature has geometry that's defined, and the value inside is the LineString enum case". If so, we're grabbing the points in that enum case and calling them `pts`. `ref` means to immutably borrow -- we don't need to "move" or "take ownership" of the data in there, just have a read-only pointer to it.

The array lookup stuff below is based on the type that we extracted: https://docs.rs/geojson/latest/geojson/type.LineStringType.html

Reply:
https://doc.rust-lang.org/rust-by-example/flow_control/match/destructuring/destructure_enum.html is a better reference for `match` / `if let`